runtime/interpreter/decode.go

```diff
@@ -1377,41 +1660,45 @@ func (d *DecoderV4) decodeRestrictedStaticType(v interface{}) (StaticType, error
 	}, nil
 }
 
-func (d *DecoderV4) decodeType(v interface{}) (TypeValue, error) {
-	encoded, ok := v.(cborArray)
+func (d *DecoderV4) decodeType() (TypeValue, error) {
 	const expectedLength = encodedTypeValueTypeLength
-	if !ok || len(encoded) != expectedLength {
-		return TypeValue{}, fmt.Errorf("invalid type encoding: expected [%d]interface{}, got %T",
+
+	arraySize, err := d.decoder.DecodeArrayHead()
+
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return TypeValue{}, fmt.Errorf("invalid type encoding: expected [%d]interface{}, got %s",
+				expectedLength,
+				e.ActualType.String(),
+			)
+		}
+		return TypeValue{}, err
+	}
+
+	if arraySize != expectedLength {
+		return TypeValue{}, fmt.Errorf("invalid type encoding: expected [%d]interface{}, got [%d]interface{}",
 			expectedLength,
-			v,
+			arraySize,
 		)
 	}
 
-	var staticType StaticType
-
-	staticTypeField := encoded[encodedTypeValueTypeFieldKey]
-	if staticTypeField != nil {
-		var err error
-		staticType, err = d.decodeStaticType(staticTypeField)
-		if err != nil {
-			return TypeValue{}, fmt.Errorf("invalid type encoding: %w", err)
-		}
+	// Decode type at array index encodedTypeValueTypeFieldKey
+	staticType, err := d.decodeStaticType()
+	if err != nil {
+		return TypeValue{}, fmt.Errorf("invalid type encoding: %w", err)
```

Comment:
As mentioned above, the static type is optional here

Reply:
I updated PR to premove optional static type decoding from decodeStaticType(). That is now done in 3 places which need optional static type decoding.